flit_core/flit_core/wheel.py

```diff
@@ -101,7 +101,7 @@ def wheel_filename(self):
         tag = ('py2.' if self.metadata.supports_py2 else '') + 'py3-none-any'
         return '{}-{}-{}.whl'.format(
                 re.sub(r"[^\w\d.]+", "_", self.metadata.name, flags=re.UNICODE),
-                re.sub(r"[^\w\d.]+", "_", self.metadata.version, flags=re.UNICODE),
+                re.sub(r"[^\w\d\+.]+", "_", self.metadata.version, flags=re.UNICODE),
                 tag)
```

Comment:
In compliance with the [recently changed spec](https://packaging.python.org/specifications/binary-distribution-format/#escaping-and-unicode), this should be the correct:

```suggestion
         # See https://packaging.python.org/specifications/binary-distribution-format/#escaping-and-unicode
         assert '-' not in self.metadata.version, 'Normalized versions can’t have dashes'
         assert '-' not in tag, 'Tags can’t have dashes'
         return '{}-{}-{}.whl'.format(
                re.sub(r"[-_.]+", "_", self.metadata.name, flags=re.UNICODE),
                self.metadata.version,
                tag)
```

Reply:
Where does `self.metadata.version` come from? Dashes are allowed in Core metadata (the field used in the `{name}-{version}.dist-info/METADATA` file), so the assertion may not be correct depending on how Flit implements this.

Reply:
`Metadata` is created via

https://github.com/takluyver/flit/blob/8f93601d05abe00c833663df541200114dbf3ad1/flit_core/flit_core/common.py#L364-L368

where `get_info_from_module` does

https://github.com/takluyver/flit/blob/8f93601d05abe00c833663df541200114dbf3ad1/flit_core/flit_core/common.py#L180

and `check_version` calls

https://github.com/takluyver/flit/blob/8f93601d05abe00c833663df541200114dbf3ad1/flit_core/flit_core/common.py#L205

Reply:
I think @flying-sheep's suggestion is correct - so long as our implementation of version normalisation is correct (and I've read PEP 440 correctly!), there should be no dashes in `version`.

I'd also like to check when constructing the filename that the compatibility tag has exactly two dashes (`{python tag}-{abi tag}-{platform tag}`), to avoid any nasty surprises. But that check can be added later.